src/WCSimAmBeGen.cc

```diff
@@ -0,0 +1,127 @@
+#include "WCSimAmBeGen.hh"
+#include "WCSimPrimaryGeneratorMessenger.hh"
+#include "WCSimPrimaryGeneratorAction.hh"
+
+#include "G4ParticleGun.hh"
+#include "G4PhysicalConstants.hh"
+#include "G4RandomDirection.hh"
+#include "G4SystemOfUnits.hh"
+#include "G4SPSEneDistribution.hh"
+#include "G4SPSRandomGenerator.hh"
+#include "G4ParticleTable.hh"
+#include "G4Neutron.hh"
+#include "G4Gamma.hh"
+
+#include <random>
+
+using namespace std;
+
+WCSimAmBeGen::WCSimAmBeGen(){
+  // Initialise
+  this->Initialise();
+  wcsimdir = string(getenv("WCSIMDIR"))+"/data/";
+}
+
+WCSimAmBeGen::~WCSimAmBeGen(){
+  
+  // This needed to be deleted
+  delete myAmBeGun;
+  delete nEnergyDist;
+  delete rGen;
+}
+
+void WCSimAmBeGen::Initialise(){
+    nEnergyDist = new G4SPSEneDistribution();
+    rGen        = new G4SPSRandomGenerator();
+    myAmBeGun   = new G4ParticleGun();
+}
+
+G4double WCSimAmBeGen::GammaEnergy(){
+    
+    // Initialise Geant4 random number generator
+    G4Random::setTheEngine(new CLHEP::RanecuEngine);
+
+    // Define arrays
+    G4double probabilities[] = {0.26, 0.65, 0.08};
+    G4double energies[] = {0.0, 4.4 * MeV, 7.7 * MeV};
+
+    // Generation of the random number
+    G4double random_number = G4UniformRand();
+
+    // Select the energy taking into account the probabilities
+    G4double cumulative_probability = 0.0;
+    G4int selected_energy_index = -1;
+
+    for (G4int j = 0; j < sizeof(probabilities) / sizeof(probabilities[0]); ++j) {
+        cumulative_probability += probabilities[j];
+        if (random_number <= cumulative_probability) {
+            selected_energy_index = j;
+            break;
+        }
+    }
+
+    gEnergy = energies[selected_energy_index];
+
+    return gEnergy; 
+}
+
+G4double WCSimAmBeGen::NeutronEnergy(G4double gEnergy){
+    // Depending on the gEnergy, we load the correspondent neutron energy spectrum
+    nEnergyDist->SetEnergyDisType("Arb");
+    
+    if (gEnergy == 0.0) {
+      gs_path = wcsimdir + "ground_state_spectrum.txt";
+      nEnergyDist->ArbEnergyHistoFile(gs_path);
+      nEnergyDist->SetBiasRndm(rGen);
+      nEnergyDist->ArbInterpolate("Lin");
+      nEnergy = nEnergyDist->GenerateOne(G4Neutron::Definition());
+    }
+    else if (gEnergy == 4.4) {
```

Comment:
hmm direct double comparison is dangerous

Reply:
What about this change:
-    else if (gEnergy == 4.4) {
+    else if (std::abs(gEnergy - 4.4) < epsilon) {

Reply:
Yeah, or even 
```C++
else if (std::abs(gEnergy - 4.4) < 1E-6) {
```